gnrc_netif: assume flag settings on compile time

miri64 opened the pull request:
This PR optimizes a lot of code out, by making the flag checker functions of `gnrc_netif` default to static values based on certain compile-time configurations. This leads to some impressive ROM changes (see table below; `✗` means the board did not compile in #7925 or is black listed, but might be with this PR, if there is no information on the board it did not build for any configuration, even with this PR).

* For host configuration I used the `gcoap` example
* For router configuration I used the `gnrc_networking` example
* For border router configuration I used the `gnrc_border_router` example.

| Board | Host | Router | Border Router |
| :---- | ---: | -----: | ------------: |
| airfy-beacon | -108 | -576 | ✗ |
| arduino-due | -60 | -404 | -1024 |
| arduino-duemilanove | -84 | -708 | -1660 |
| arduino-mega2560 | -84 | -706 | -1662 |
| arduino-mkr1000 | -64 | -484 | -1180 |
| arduino-mkrzero | -64 | -484 | -1180 |
| arduino-uno | -84 | -708 | -1660 |
| arduino-zero | -64 | -484 | -1180 |
| avsextrem | -88 | -648 | -1632 |
| b-l072z-lrwan1 | -64 | -484 | ✗ |
| calliope-mini | -64 | -484 | ✗ |
| cc2538dk | -460 | -2380 | -1024 |
| cc2650-launchpad | -60 | -404 | ✗ |
| cc2650stk | -60 | -404 | ✗ |
| ek-lm4f120xl | -48 | -400 | -1040 |
| f4vi1 | -64 | -400 | -1040 |
| fox | -460 | -2380 | -1024 |
| frdm-k22f | -64 | -416 | -1024 |
| frdm-k64f | -64 | -400 | -1040 |
| iotlab-a8-m3 | -460 | -2380 | -1024 |
| iotlab-m3 | -460 | -2380 | -1024 |
| limifrog-v1 | -60 | -404 | -1024 |
| maple-mini | -60 | -404 | ✗ |
| mbed_lpc1768 | -60 | -404 | -1024 |
| microbit | -108 | -576 | ✗ |
| mips-malta | -116 | -652 | ✗ |
| msba2 | -720 | -3976 | -1632 |
| msbiot | -64 | -400 | -1024 |
| mulle | -464 | -2384 | -1024 |
| native | -816 | -1848 | ✗ |
| nrf51dongle | -108 | -576 | ✗ |
| nrf52840dk | -112 | -496 | -1024 |
| nrf52dk | -304 | -1376 | -1024 |
| nrf6310 | -108 | -576 | ✗ |
| nucleo-f070 | -64 | -484 | ✗ |
| nucleo-f072 | -64 | -484 | ✗ |
| nucleo-f091 | -64 | -484 | -1180 |
| nucleo-f103 | -60 | -404 | ✗ |
| nucleo-f302 | -64 | -400 | ✗ |
| nucleo-f303 | -64 | -400 | -1040 |
| nucleo-f401 | -64 | -400 | -1024 |
| nucleo-f410 | -64 | -400 | -1024 |
| nucleo-f411 | -64 | -400 | -1024 |
| nucleo-f446 | -64 | -400 | -1024 |
| nucleo-l073 | -64 | -484 | ✗ |
| nucleo-l152 | -60 | -404 | -1024 |
| nucleo-l476 | -64 | -400 | -1040 |
| nucleo144-f207 | -60 | -404 | -1024 |
| nucleo144-f303 | -64 | -400 | -1024 |
| nucleo144-f412 | -48 | -416 | -1024 |
| nucleo144-f413 | -64 | -400 | -1040 |
| nucleo144-f429 | -48 | -416 | -1024 |
| nucleo144-f446 | -48 | -416 | -1024 |
| nucleo144-f722 | -64 | -400 | -1024 |
| nucleo144-f746 | -64 | -416 | -1040 |
| nucleo144-f767 | -64 | -416 | -1040 |
| nucleo32-f303 | -64 | -400 | ✗ |
| nucleo32-l432 | -48 | -400 | -1024 |
| nz32-sc151 | -60 | -404 | -1024 |
| opencm904 | -60 | -404 | ✗ |
| openmote-cc2538 | -460 | -2380 | -1024 |
| pba-d-01-kw2x | -464 | -2384 | -1024 |
| pca10000 | -108 | -576 | ✗ |
| pca10005 | -108 | -576 | ✗ |
| pic32-clicker | -112 | -656 | ✗ |
| pic32-wifire | -84 | -476 | ✗ |
| remote-pa | -460 | -2380 | -1024 |
| remote-reva | -460 | -2380 | -1024 |
| remote-revb | -460 | -2380 | -1024 |
| samd21-xpro | -64 | -484 | -1180 |
| saml21-xpro | -64 | -484 | -1180 |
| samr21-xpro | -504 | -2684 | -1184 |
| seeeduino_arch-pro | -60 | -404 | -1024 |
| slwstk6220a | -64 | -416 | -1024 |
| sodaq-autonomo | -64 | -484 | -1180 |
| sodaq-explorer | -64 | -484 | -1180 |
| spark-core | -60 | -404 | ✗ |
| stm32f3discovery | -48 | -416 | -1024 |
| stm32f4discovery | -64 | -400 | -1024 |
| stm32f7discovery | -64 | -400 | -1024 |
| udoo | -60 | -404 | -1024 |
| waspmote-pro | -84 | -708 | -1660 |
| yunjia-nrf51822 | -108 | -576 | ✗ |

gebart commented:
Impressive savings.
Does it affect execution time much if those functions are modified to be normal functions instead of `static inline`?

miri64 commented:
> Impressive savings.
Does it affect execution time much if those functions are modified to be normal functions instead of static inline?

Did not evaluate that, but I would guess that the optimization wouldn't be quite that impressive than, because the values wouldn't be `const` anymore, right?

gebart commented:
@miri64 I did not mean to question this PR, I think this implementation is great. I was only wondering if we could improve the other case as well. If the very large size savings were because of many instantiations of the inline function which would be saved by making it a regular function with only one instance.

miri64 commented:
Ah... I've misread your comment and thought you wanted to make them static inline instead of macros ^^".

I don't think this optimization is so effective is because of the number of instantiations of the inline functions (after all, it is just a flag check) but because a lot of else-branches are dropped by the optimizer, since many if-conditions become constant with this PR.

bergzand commented:
Really nice, will take a look this evening.

> but because a lot of else-branches that are dropped by the optimizer, since many if-conditions become constant with this PR.

I was wondering about this since those optimizations on themselves look to small to save this amount of code :)

miri64 commented:
Closing of #7925 seem to have caused this...

smlng commented:
interesting Murdock errors, one way to solve them would be:

```
 static inline bool gnrc_netif_is_rtr(const gnrc_netif_t *netif)
 {
#if defined(MODULE_IPV6_ROUTER) || defined(DOXYGEN)
    return (netif->flags & GNRC_NETIF_FLAGS_IPV6_FORWARDING);
#else
    (void)netif;
    return false;
#endif
 }
```

though I haven't checked if changing all functions to that style achieves the same size savings?!

miri64 commented:
Nah, that would negate the effect of the optimization. I'll `(void)netif` before the "call" ;-).

bergzand commented:
Same construction [here](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/network_layer/ipv6/nib/_nib-router.c#L163)

bergzand commented:
I find it curious how for some platforms the gains for the border router are larger than the router (nucleo-*), where for others it is reversed (iotlab-m3).

Any recommended approach for testing this? Or just test the 3 cases of 6lbr, 6lr and 6ln?

miri64 commented:
> I find it curious how for some platforms the gains for the border router are larger than the router (nucleo-*), where for others it is reversed (iotlab-m3).

I was wondering about this as well, but don't really have an explanation. :-/

> Any recommended approach for testing this? Or just test the 3 cases of 6lbr, 6lr and 6ln?

I think the last one is the best way to cover all bases.

miri64 commented:
(I'm also still somewhat surprised that the biggest optimization is in the border router, though it has the most features => all this optimizations hould evaluate to the real function....)

miri64 commented:
Since I changed some lines and checks, maybe it is a good idea to redo the table. Will do.

miri64 commented:
>> I find it curious how for some platforms the gains for the border router are larger than the router (nucleo-*), where for others it is reversed (iotlab-m3).
>
> I was wondering about this as well, but don't really have an explanation. :-/

@bergzand While waiting for the results, I got an idea what might be the cause: The `iotlab-m3` e.g. has an IEEE 802.15.4 radio, causing the build system to include 6LoWPAN, thus to compile in 6LN/6LR/6LBR configuration. `nucleo-*` however doesn't have a radio, thus classic IPv6 configuration is assumed (i.e. completely different image).

miri64 commented:
Finally done (and rebased and squashed to current master). Minor change to the notification; a ✔ now marks if an application can now be linked with this PR (but previously not):

| Board | Host | Router | Border Router |
| :---- | ---: | -----: | ------------: |
| airfy-beacon | -116 | -592 | ✘ |
| arduino-due | -72 | -416 | -1024 |
| arduino-duemilanove | -86 | -656 | -1700 |
| arduino-mega2560 | -86 | -658 | -1702 |
| arduino-mkr1000 | -72 | -500 | -1172 |
| arduino-mkrzero | -72 | -500 | -1172 |
| arduino-uno | -86 | -656 | -1700 |
| arduino-zero | -72 | -500 | -1172 |
| avsextrem | -88 | -656 | -1640 |
| b-l072z-lrwan1 | -72 | -500 | ✘ |
| bluepill | -68 | -416 | ✘ |
| calliope-mini | -72 | -500 | ✘ |
| cc2538dk | -1052 | -1888 | -1028 |
| cc2650-launchpad | -68 | -416 | ✘ |
| cc2650stk | -68 | -416 | ✘ |
| ek-lm4f120xl | -64 | -416 | -1040 |
| f4vi1 | -80 | -416 | -1040 |
| feather-m0 | -72 | -500 | -1172 |
| fox | -1052 | -1888 | -1028 |
| frdm-k22f | -64 | -416 | -1024 |
| frdm-k64f | -64 | -416 | -1024 |
| ikea-tradfri | -64 | -416 | -1040 |
| iotlab-a8-m3 | -1052 | -1888 | -1028 |
| iotlab-m3 | -1052 | -1888 | -1028 |
| limifrog-v1 | -68 | -416 | -1028 |
| maple-mini | -68 | -416 | ✘ |
| mbed_lpc1768 | -68 | -416 | -1028 |
| microbit | -116 | -592 | ✘ |
| mips-malta | -108 | -644 | ✘ |
| msba2 | -1664 | -3000 | -1640 |
| msbiot | -80 | -416 | -1040 |
| mulle | -1056 | -1888 | -1040 |
| native | -816 | -1848 | -6680 |
| nrf51dongle | -116 | -592 | ✘ |
| nrf52840dk | -80 | -416 | -1024 |
| nrf52dk | -224 | -592 | -1040 |
| nrf6310 | -116 | -592 | ✘ |
| nucleo-f070 | -72 | -500 | ✘ |
| nucleo-f072 | -72 | -500 | ✘ |
| nucleo-f091 | -72 | -500 | -1172 |
| nucleo-f103 | -68 | -416 | ✘ |
| nucleo-f302 | -64 | -416 | ✔ |
| nucleo-f303 | -64 | -416 | -1024 |
| nucleo-f334 | -80 | -416 | ✘ |
| nucleo-f401 | -64 | -416 | -1024 |
| nucleo-f410 | -64 | -416 | -1024 |
| nucleo-f411 | -64 | -416 | -1024 |
| nucleo-f446 | -80 | -416 | -1040 |
| nucleo-l073 | -72 | -500 | ✘ |
| nucleo-l152 | -68 | -416 | -1028 |
| nucleo-l476 | -80 | -416 | -1024 |
| nucleo144-f207 | -68 | -416 | -1028 |
| nucleo144-f303 | -80 | -416 | -1040 |
| nucleo144-f412 | -64 | -416 | -1040 |
| nucleo144-f413 | -80 | -416 | -1040 |
| nucleo144-f429 | -64 | -416 | -1040 |
| nucleo144-f446 | -64 | -416 | -1040 |
| nucleo144-f722 | -64 | -416 | -1040 |
| nucleo144-f746 | -64 | -416 | -1040 |
| nucleo144-f767 | -64 | -416 | -1040 |
| nucleo32-f303 | -64 | -416 | ✘ |
| nucleo32-l432 | -64 | -416 | -1040 |
| nz32-sc151 | -68 | -416 | -1028 |
| opencm904 | -68 | -416 | ✘ |
| openmote-cc2538 | -1052 | -1888 | -1028 |
| pba-d-01-kw2x | -1056 | -1888 | -1040 |
| pic32-clicker | -112 | -656 | ✘ |
| pic32-wifire | -76 | -472 | ✘ |
| remote-pa | -1052 | -1888 | -1028 |
| remote-reva | -1052 | -1888 | -1028 |
| remote-revb | -1052 | -1888 | -1028 |
| samd21-xpro | -72 | -500 | -1172 |
| saml21-xpro | -72 | -500 | -1172 |
| samr21-xpro | -1208 | -2124 | -1176 |
| seeeduino_arch-pro | -68 | -416 | -1028 |
| sltb001a | -64 | -432 | -1040 |
| slwstk6220a | -64 | -416 | -1024 |
| sodaq-autonomo | -72 | -500 | -1172 |
| sodaq-explorer | -72 | -500 | -1172 |
| spark-core | -68 | -416 | ✘ |
| stm32f3discovery | -64 | -416 | -1024 |
| stm32f4discovery | -64 | -416 | -1024 |
| stm32f7discovery | -64 | -416 | -1040 |
| udoo | -72 | -416 | -1024 |
| waspmote-pro | -86 | -656 | -1700 |
| yunjia-nrf51822 | -116 | -592 | ✘ |

bergzand commented:
Testing this on three nodes:

- 6lbr: gnrc_border_router on nucleo-f446 with mrf24j40.
- 6lr: gnrc_networking on nucleo-f401 with mrf24j40.
- 6ln: gcoap example on nucleo-f401 with mrf24j40.

The gnrc_networking seems to hang after with gnrc_rpl enabled. Without gnrc_rpl, the RTR flag is not set on the interface, while it is set before these commits

miri64 commented on sys/include/net/gnrc/netif/internal.h:
```diff
@@ -300,32 +306,45 @@ int gnrc_netif_ipv6_get_iid(gnrc_netif_t *netif, eui64_t *eui64);
  * @return  true, if the interface represents a router
  * @return  false, if the interface does not represent a router
  */
+#if defined(MODULE_IPV6_ROUTER) || defined(DOXYGEN)
```
Welp… that module doesn't exist (it's supposed to be `MODULE_GNRC_IPV6_ROUTER`)... Will fix.

miri64 commented:
I guess the ROM optimization won't be that impressive anymore :-/

bergzand commented on sys/include/net/gnrc/netif/internal.h:
```diff
@@ -349,16 +375,25 @@ bool gnrc_netif_is_6ln(const gnrc_netif_t *netif);
  * @return  true, if the interface represents a 6LR
  * @return  false, if the interface does not represent a 6LR
  */
+#if (GNRC_IPV6_NIB_6LR && \
```
I think this should be `GNRC_IPV6_NIB_CONF_6LR`. This might explain why neighbor address registration is broken.

miri64 replied:
I think I might have been drunk, when I implemented that 😅

gebart commented:
Idea for a static check tool: greping the sources for any conditionals based on modules and parsing whether any part of the tree provides that module

bergzand commented:
> Idea for a static check tool: greping the sources for any conditionals based on modules and parsing whether any part of the tree provides that module

@gebart I had the same idea this morning, I created #8284 to track it.

@miri64 Looks good and functional now. Verified it against the branch ancestor (or whatever the commit this branch is branched from is called). Behaviour of the three nodes looks identical to me.

ACK, Please squash

miri64 commented:
Squashed
